arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Bracket.kt

```diff
@@ -231,31 +206,32 @@ public suspend inline fun <A, B> bracketCase(
   use: suspend (A) -> B,
   crossinline release: suspend (A, ExitCase) -> Unit
 ): B {
-  val acquired = withContext(NonCancellable) {
-    acquire()
-  }
-
-  val res = try {
-    use(acquired)
-  } catch (e: CancellationException) {
-    runReleaseAndRethrow(e) { release(acquired, ExitCase.Cancelled(e)) }
-  } catch (t: Throwable) {
-    runReleaseAndRethrow(t.nonFatalOrThrow()) { release(acquired, ExitCase.Failure(t.nonFatalOrThrow())) }
-  }
-
-  withContext(NonCancellable) { release(acquired, ExitCase.Completed) }
-
-  return res
+  val acquired = withContext(NonCancellable) { acquire() }
+  return guaranteeCase({ use(acquired) }) { release(acquired, it) }
 }
 
 @PublishedApi
-internal suspend inline fun runReleaseAndRethrow(original: Throwable, crossinline f: suspend () -> Unit): Nothing {
+internal suspend inline fun runReleaseAndRethrow(original: Throwable?, crossinline f: suspend () -> Unit) {
```

Comment:
Why has this changed to a nullable `Throwable`? In which situations would `null` be the value here?

Reply:
In successful completion. This method basically acts as `withContext(NonCancellable) { f()` if given a null throwable (which is the "happy path" case).

Reply:
But why before it didn't have it? Was the handling of the happy path done somewhere else? If so, I would prefer to keep this without the `?` and inline the corresponding implementation for the `null` cases.

Reply:
Before, this was used directly when a throwable was `caught`. Now, we instead use `finalizeCase`. The point is to unify where the release happens one place. Actually, looking at it now, I think this method can be inlined and simplified. One second...

Reply:
Changed!